### DISCUSSION TITLE
How to get individual elements from a smiles string

### DISCUSSION THREAD

**Discussion by pv-is-nrt** [rdkit/rdkit#5690] [Q&A]
In Python, I am trying to break a SMILES string into a list of SMILES characters.
For example, I want to convert this string CC(Cl)c1ccn(C)c1 into this list ['C', 'C', '(', 'Cl', ')', 'c', '1', 'c', 'c', 'n', '(', 'C', ')', 'c', '1']. This is not as straightforward as simply getting the characters from the string: occurrence of a smaller case alphabet could either mean an element denoted by more than one character (like Cl for Chlorine) or indicate that the element is part of an aromatic ring (like n for Nitrogen). In the former case, I want the character to be included with the element's name first letter and in the latter case I want the character to be identified as an element.
Before I write a parsing algorithm to accomplish this, which I think I can but would be less than ideal because I might miss a rule here and there (number > 9 is an additional thing that would need to be tackled), I wanted to ask if rdkit already has a method to do this kind of deconstruction of the SMILES string.

**Comment by vfscalfani** [comment]
Hi @pv-is-nrt.

Here are a few pieces to get started...

You can get the atom symbols and determine if the atom is considered aromatic by parsing the SMILES, then looping over the atoms like this:

```
m = Chem.MolFromSmiles("CC(Cl)c1ccn(C)c1")
for i in range(m.GetNumAtoms()):
    print(m.GetAtomWithIdx(i).GetSymbol(), m.GetAtomWithIdx(i).GetIsAromatic())
```

```
C False
C False
Cl False
C True
C True
C True
N True
C False
C True
```

I'm assuming you could then use the True/False to lowercase element symbols as needed. I'm not sure how you would reconstruct the branching/ring number symbols...maybe start with getting the connectivity of atoms:

```
for bond in m.GetBonds():
    print(bond.GetBeginAtomIdx(),bond.GetEndAtomIdx())
```

```
0 1
1 2
1 3
3 4
4 5
5 6
6 7
6 8
8 3
```

**Comment by DocMinus** [comment]
Here is a solution I wrote without the use of Rdkit, returns a elemental count; without Hs: 

```  
import re

smiles = "CC(Cl)c1ccn(C)c1"

pattern = "Si|Ti|Al|Zn|Pd|Pt|Br?|Cl?|N|O|S|P|F|I|B|b|c|n|o|s|p" 
regex = re.compile(pattern)
elements = [token for token in regex.findall(smiles)]
ele_length = len(elements)
lowercase_pattern = "b|c|n|o|s|p"
regex_low = re.compile(lowercase_pattern)
for i in range(ele_length):
    if regex_low.findall(elements[i]):
        elements[i] = elements[i].upper()
element_count = [elements.count(ele) for ele in elements]
formula = dict(zip(elements, element_count)) 

print(str(formula))
```  
Output:
```  
{'C': 7, 'Cl': 1, 'N': 1}
```
